pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/CompoundDataBuffer.java

```diff
@@ -0,0 +1,672 @@
+
+  private final DataBuffer[] _buffers;
+  private final long[] _bufferOffsets;
+  private int _lastBufferIndex = 0;
+  private final ByteOrder _order;
+  private final long _size;
+  private final boolean _owner;
+
+  /**
+   * Creates a compound buffer from the given buffers.
+   *
+   * @param buffers The buffers that will be concatenated to form the compound buffer.
+   * @param order The byte order of the buffer. Buffers in the array that have a different byte order will be converted.
+   * @param owner Whether this buffer owns the underlying buffers. If true, the underlying buffers will be released when
+   *              this buffer is closed.
+   */
+  public CompoundDataBuffer(DataBuffer[] buffers, ByteOrder order, boolean owner) {
+    _owner = owner;
+    _buffers = buffers;
+    _bufferOffsets = new long[buffers.length];
+    _order = order;
+    long offset = 0;
+    for (int i = 0; i < buffers.length; i++) {
+      if (buffers[i].size() == 0) {
+        throw new IllegalArgumentException("Buffer at index " + i + " is empty");
+      }
+      if (buffers[i].order() != _order) {
+        buffers[i] = buffers[i].view(0, buffers[i].size(), _order);
+      }
+    }
+    for (int i = 0; i < buffers.length; i++) {
+      _bufferOffsets[i] = offset;
+      long size = buffers[i].size();
+      offset += size;
+    }
+    _size = offset;
+  }
+
+  public CompoundDataBuffer(ByteBuffer[] buffers, ByteOrder order, boolean owner) {
+    this(asDataBufferArray(buffers), order, owner);
+  }
+
+  /**
+   * Creates a compound buffer from the given buffers.
+   * @param buffers The buffers that will be concatenated to form the compound buffer.
+   * @param order The byte order of the buffer. Buffers in the list that have a different byte order will be converted.
+   * @param owner Whether this buffer owns the underlying buffers. If true, the underlying buffers will be released when
+   *              this buffer is closed.
+   */
+  public CompoundDataBuffer(List<DataBuffer> buffers, ByteOrder order, boolean owner) {
+    this(buffers.toArray(new DataBuffer[0]), order, owner);
+  }
+
+  private static DataBuffer[] asDataBufferArray(ByteBuffer[] buffers) {
+    DataBuffer[] result = new DataBuffer[buffers.length];
+    for (int i = 0; i < buffers.length; i++) {
+      result[i] = PinotByteBuffer.wrap(buffers[i]);
+    }
+    return result;
+  }
+
+  private int getBufferIndex(long offset) {
+    // this optimistically assumes that lookups are going to be in ascending order
+    // we don't care about concurrency here given that this is only used to speed up the lookup
+    int lastBufferIndex = _lastBufferIndex;
+    if (_bufferOffsets[lastBufferIndex] > offset) {
+      lastBufferIndex = 0;
+    }
+
+    for (int i = lastBufferIndex; i < _bufferOffsets.length; i++) {
+      if (offset < _bufferOffsets[i]) {
+        int result = i - 1;
+        _lastBufferIndex = result;
+        return result;
+      }
+    }
+    return _bufferOffsets.length - 1;
```

Comment:
Do we want to cache `_lastBufferIndex` here as well?

Reply:
Yes, it makes sense.